### ISSUE TITLE
Verify that ann is reading the entire dataset at least once

### ISSUE THREAD

**Issue by shrit** [mlpack/mlpack#2226]
Hello, 

In the actual implementation of the Artificial Neural Network, the users are using the `Train()` function to train a model. We pass a dataset into the train function,  but we are not sure that in one epoch we are passing over the entire dataset.
It would be very appropriate to print a warning telling the user that the epoch did not pass over the entire dataset, asking to modify the `maxIteration()` to be at the same size as the numbers of columns of the dataset matrix.

**Comment by bicycleman15** [comment]
Hi @zoq @birm Can I take up this issue ?

**Comment by gaurav-singh1998** [comment]
Hi, @shrit I would like to move forward with this issue. Any hint from your side on how to move forward with it would be very helpful. Thanks.

**Comment by shrit** [comment]
@gaurav-singh1998 You can choose any dataset that can be trained with neural networks, then you reduce the `maxIteration()` parameter, and monitor how the training process using `Progressbar()` callback. If you choose a small `maxIteration()` much smaller than the dataset size, you will notice that the training will stop before completing the epoch.

**Comment by gauthampkrishnan** [comment]
Can I take up this problem? Can I work on this problem?

**Comment by gaurav-singh1998** [comment]
Hey, @gauthampkrishnan I am currently working on this. Can you take another issue?

**Comment by gauthampkrishnan** [comment]
> Hey, @gauthampkrishnan I am currently working on this. Can you take another issue? Thanks.

@gaurav-singh1998 yes ofcourse all the best

**Comment by gauthampkrishnan** [comment]
Interested in a collaboration?

**Comment by gaurav-singh1998** [comment]
@gauthampkrishnan Yes, sure.

**Comment by gaurav-singh1998** [comment]
Hi, @shrit sorry for the late reply, got busy because of college mid-sem exams. I have been trying to replicate the issue you mentioned. Code regarding the same can be found [here](https://ideone.com/bZH5Xx). The dataset used was generated with the help of python interpreter and the code regarding it can be found [here](https://ideone.com/a3wbW0). I wanted to know if I am on the correct path with this and if this is correct then how to use the Progressbar() callback in
'Train()' method of Perceptron object.

**Comment by shrit** [comment]
I am not sure that perceptrons accept callbacks. You can use the code I have used in this issue https://github.com/mlpack/mlpack/issues/2073#issuecomment-549080270

**Comment by gaurav-singh1998** [comment]
Hi, @shrit can you send the dataset you used in the example?

**Comment by gaurav-singh1998** [comment]
Hey, @shrit since `MaxIterations()` is present in every optimizer, should the warning be added in every optimizer? Correct me if I am wrong.

**Comment by gaurav-singh1998** [comment]
Hey, @shrit I was going through the issue you mentioned and was able to understand the issue present. So, according to me, there has to be a warning in [ffn.hpp](https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/ann/ffn.hpp) telling the user to pass `maxIteration()` value the same as the number of columns in the dataset? Correct me if I am wrong

**Comment by shrit** [comment]
@gaurav-singh1998 You understood it right, this is exaclty what has to be done.

**Comment by gaurav-singh1998** [comment]
Thanks, @shrit for the clarification will open a PR by tomorrow :)

**Comment by gaurav-singh1998** [comment]
Hey, @shrit one final query. Can you tell what should the warning say?

**Comment by shrit** [comment]
@gaurav-singh1998 You can add something as follows: `The number of iterations is less than the size of the dataset, the optimizer might not pass over the entire dataset. Please modify the number of iterations to be at least equal to the number of columns of your dataset.`
You can modify this warning as you like, but it has to tell the user about these points.

**Comment by gaurav-singh1998** [comment]
Hi, @shrit I have incorporated the warning message in `ffn_impl.hpp`, `rnn_impl.hpp` and `brnn_impl.hpp` but if we call the `Train()` function inside a for loop the warning gets printed multiple times. Is there a way to tackle it?

**Comment by shrit** [comment]
@gaurav-singh1998 The `Train()` function itself have a loop inside (in the optimizer part), why we would have to put it inside a loop?

**Comment by gaurav-singh1998** [comment]
@shrit like in the example you used in the comment you did,
`  for(int i=0; i<1; i++)
  {
    model.Train(trainX, trainY, optimizer, ens::PrintLoss(),
                ens::ProgressBar(),
                ens::StoreBestCoordinates<arma::mat>());
    
    optimizer.ResetPolicy() = false;
  }` 
Because of this the warn message is getting multiple times.

**Comment by shrit** [comment]
@gaurav-singh1998  Exactly. I agree with you.

When I posted this issue, it was back in November. Callbacks where just begin merged in mlpack.
Later on, I understood we do not have to use outside `for` loop. In fact, it is totally possible to achieve the objective of the optimization without outside outer loop.
Before callbacks, we had to put iterations since we did not observe in real-time either the training loss or the validation loss, so you had to test on a validation set after each return of `Train()`, and write by ourself an Early Stopping locally based on the `Evaluate()`.
@zoq What do you think? Do we have to put an outside loop on the train function nowadays?

**Comment by zoq** [comment]
> When I posted this issue, it was back in November. Callbacks where just begin merged in mlpack.
> Later on, I understood we do not have to use outside for loop. In fact, it is totally possible to achieve the objective of the optimization without outside outer loop.
> Before callbacks, we had to put iterations since we did not observe in real-time either the training loss or the validation loss, so you had to test on a validation set after each return of Train(), and write by ourself an Early Stopping locally based on the Evaluate().
> @zoq What do you think? Do we have to put an outside loop on the train function nowadays?

You are absolutely correct, the callbacks should make the for loop obsolete.

**Comment by gaurav-singh1998** [comment]
Hi, @shrit since this issue has been solved. I think we should go ahead and close it. Thanks.
